Cloud SMTP Whitelabeling

fraserdrops opened the pull request:
Custom SMTP configuration is available for self-hosted instances, but on Cloud Metabase manages SMTP for the user. However, users want to be able to customize the From Address so that emails come from their own domain rather than from noreply at metabase.com.

## Self-hosted instances
This is largely unchanged. Users can still configure custom SMTP like before.

Send test email button is now on the main email settings page, this is true for all cases not just self-hosted
<img width="600" alt="Screenshot 2025-06-25 at 8 05 38 PM" src="https://github.com/user-attachments/assets/66209e85-abe7-4418-bb1e-fabd09a55581" />

## Starter cloud instances
- still cannot configure custom SMTP
- the From Address appears locked with a message explaining it is not configurable because they don't have the pro-plan which would allow them to set up custom SMTP
- Gets shown an upsell for email whitelabeling
<img width="600" alt="Screenshot 2025-06-25 at 8 09 08 PM" src="https://github.com/user-attachments/assets/a75622e4-f7a4-49ed-8846-59031e174741" />

## Pro-cloud instances
- The From Address starts off locked when custom SMTP is not configured
<img width="600" alt="Screenshot 2025-06-25 at 8 12 48 PM" src="https://github.com/user-attachments/assets/a4b03e1f-05ef-4776-9f84-4809a11d7d4e" />

- An SMTP configuration card is visible
- Once custom SMTP is configured, the card allows the user to switch between the custom config and 'Metabased Managed'
- Once custom SMTP is configured, the From Address is now enabled
<img width="600" alt="Screenshot 2025-06-25 at 8 13 34 PM" src="https://github.com/user-attachments/assets/768f21ec-7cad-417e-bc36-c56e8775c7f6" />


## Testing
The new dev tokens can be used for testing.

An annoying thing is that in a real cloud instance a bunch of settings get set by env var, including some email ones. Without this, the email page won't display all the fields because it doesn't see that email is configured. A couple of ways around this:
1. Start off with a self-hosted instance, configure smtp there, then upgrade the licence.
2. In `EmailSettingsPage.tsx` component, add `|| true` to `  const isEmailConfigured = settingValues?.["email-configured?"] || true;`

## Still TODO
- When custom SMTP gets configured, it should automatically be enabled. Currently it stays 'Metabase Managed' until the user manually changes it
- Testing cloud custom SMTP requires a secure SMTP server. Our e2e tests only have a simple maildev server which is not secure. We need to add this secure server to fully e2e test

nvoxland committed:
Added /api/email/cloud API and backing settings

nvoxland committed:
Added /api/email/cloud API and backing settings

fraserdrops committed:
Create cloud version of smtp form

fraserdrops committed:
Create cloud version of smtp form

nvoxland committed:
Merge branch 'cloud-smtp' into feature-cloud-smtp-whitelabeling

# Conflicts:
#	test/metabase/channel/settings_test.clj

nvoxland committed:
Added cloud-smtp-enabled setting to PUT /api/cloud

nvoxland committed:
Added cloud-smtp-enabled setting to PUT /api/cloud

fraserdrops committed:
Add chip forms for security and port fields

fraserdrops committed:
Add chip forms for security and port fields

fraserdrops committed:
Add EmailFromAddressWidget

fraserdrops committed:
Add EmailFromAddressWidget

fraserdrops committed:
Fix form validation

fraserdrops committed:
Fix form validation

fraserdrops committed:
Connect connection card to cloud-smtp-enabled? setting

nvoxland committed:
Added cloud-custom-smtp feature flag
Fixed frontend SMTP port from 456 to 465

nvoxland committed:
Added cloud-custom-smtp feature flag
Fixed frontend SMTP port from 456 to 465

nvoxland committed:
Added delete API ADM-993

nvoxland committed:
Added delete API ADM-993

nvoxland committed:
Email sending should respect the cloud-smtp-enabled? setting ADM-994

nvoxland committed:
Email sending should respect the cloud-smtp-enabled? setting ADM-994

nvoxland committed:
Remove cloud-smtp-enabled setting from /api/email/cloud API

nvoxland committed:
Remove cloud-smtp-enabled setting from /api/email/cloud API

nvoxland committed:
Only allow setting cloud-smtp-enabled? when cloud-email-smtp-host is set

nvoxland committed:
Only allow setting cloud-smtp-enabled? when cloud-email-smtp-host is set

nvoxland committed:
Updated tests

nvoxland committed:
Updated tests

fraserdrops committed:
Tidy upsell

fraserdrops committed:
Fix syntax error

nvoxland committed:
Merge remote-tracking branch 'origin/feature-cloud-smtp-whitelabeling' into feature-cloud-smtp-whitelabeling

# Conflicts:
#	frontend/src/metabase-types/api/mocks/settings.ts
#	frontend/src/metabase-types/api/settings.ts
#	frontend/src/metabase/admin/settings/components/Email/CloudSMTPConnectionCard.tsx
#	frontend/src/metabase/admin/settings/components/SettingsPages/EmailSettingsPage.tsx
#	frontend/src/metabase/api/email.ts
#	src/metabase/channel/settings.clj
#	test/metabase/channel/settings_test.clj

nvoxland committed:
Renamed custom-smtp-enabled? to custom-smtp-enabled

fraserdrops committed:
Add FromAddress spec

fraserdrops committed:
Merge branch 'feature-cloud-smtp-whitelabeling' of https://github.com/metabase/metabase into feature-cloud-smtp-whitelabeling

nvoxland committed:
Require is-hosted? for using custom-smtp

nvoxland committed:
Added split off /api/email/cloud/test API

nvoxland committed:
Added split off /api/email/cloud/test API

nvoxland committed:
Removed unneeded cloud-* settings

nvoxland committed:
Removed unused from/reploy-to settings from PUT /api/email and PUT /api/email/cloud

fraserdrops committed:
Remove unneeded isHosted check

fraserdrops committed:
Rework connection cards

fraserdrops committed:
Use cloud version of sendTestEmail

fraserdrops committed:
Add test for CloudSMTPConnectionForm

fraserdrops committed:
Merge branch 'master' into feature-cloud-smtp-whitelabeling

nvoxland committed:
Removed separate :cloud vs. :standard email sending

fraserdrops committed:
Add tracking events

fraserdrops committed:
Improve error handling in forms

fraserdrops committed:
Update EmailFromAddressWidget

fraserdrops committed:
Merge branch 'master' into feature-cloud-smtp-whitelabeling

fraserdrops committed:
Fix import

fraserdrops committed:
Move send email button to main email settings page

fraserdrops committed:
Merge branch 'master' into feature-cloud-smtp-whitelabeling

fraserdrops committed:
Fix existing email tests

fraserdrops committed:
Add tests for starter instance

fraserdrops committed:
Add pro-cloud tests

fraserdrops committed:
Add to mock settings

fraserdrops commented:
@cubic-dev-ai review this PR

fraserdrops commented on frontend/src/metabase/admin/settings/components/Email/CloudSMTPConnectionCard.module.css:
```diff
@@ -0,0 +1,23 @@
+}
+
+.indicator {
+  border: 1px solid gray;
```
I need to change this colour, but the bigger question is why it's not displaying by default. Something is happening with the Mantine component styles not displaying correctly. I copied code directly from the Mantine example and it doesn't work

fraserdrops committed:
AI overlord fixes

fraserdrops committed:
Add cypress tests for pro-cloud instance

nvoxland committed:
Update PUT /api/email and /api/email/cloud to be more of a PUT than a PATCH and also sets cloud-smtp-enabled

fraserdrops committed:
Turn upsell banner into pill

fraserdrops committed:
Add loading spinner when submitting

fraserdrops committed:
Disable clear button while submitting

iethree commented on frontend/src/metabase/admin/settings/components/Email/SMTPConnectionCard.tsx:
```diff
@@ -1,38 +1,46 @@
+    );
+  }
+
+  if (isHosted && hasCloudSMTPFeature) {
+    return (
+      <>
+        <CloudSMTPConnectionCard
+          onOpenCloudSMTPModal={() => {
+            openModal();
+            trackSMTPSetupClick({ eventDetail: "cloud" });
+          }}
+        />
+        {showModal && <CloudSMTPConnectionForm onClose={closeModal} />}
+      </>
+    );
+  }
```
I think `CloudSMTPConnectionCard` needs to be an enterprise plugin

nvoxland committed:
Added maildev-ssl to e2e docker-compose.yml

nvoxland committed:
Added maildev-ssl to e2e docker-compose.yml

nvoxland committed:
Added maildev-ssl to e2e docker-compose.yml

fraserdrops committed:
Tidy tests

nvoxland committed:
Added dev.keystore.jks for dev instances to always use

nvoxland committed:
Merge remote-tracking branch 'origin/feature-cloud-smtp-whitelabeling' into feature-cloud-smtp-whitelabeling

nvoxland committed:
Merge remote-tracking branch 'origin/master' into feature-cloud-smtp-whitelabeling

# Conflicts:
#	.github/workflows/e2e-test.yml

nvoxland committed:
Start maildev-ssl for CI

fraserdrops committed:
Add to e2e tests

fraserdrops committed:
Add snowplow tracking

nvoxland committed:
Use mail.smtps.ssl.trust=* instead of a custom trust store

nvoxland committed:
Fixed linting errors

iethree commented on frontend/src/metabase/admin/settings/components/Email/CloudSMTPConnectionCard.tsx:
```diff
@@ -0,0 +1,93 @@
+        onChange={handleChange}
+      >
+        <Radio.Card radius="md" value="metabase" className={S.RadioCardRoot}>
+          <Box p="lg">
+            <Group justify="space-between">
```
why do we need this Box ? can't we just put the padding on the Group?

iethree commented on frontend/src/metabase/admin/settings/components/Email/CloudSMTPConnectionForm.tsx:
```diff
@@ -0,0 +1,293 @@
+                  settingKey="cloud-email-smtp-port"
+                  settingDetails={settingsDetails?.["cloud-email-smtp-port"]}
+                >
+                  <FormChipGroup
+                    name="cloud-email-smtp-port"
+                    label={t`SMTP Port`}
+                    groupProps={{ mt: "0.5rem" }}
+                  >
+                    <Chip value={"465"} variant="brand">
+                      465
+                    </Chip>
+                    <Chip value={"587"} variant="brand">
+                      587
+                    </Chip>
+
+                    <Chip value={"2525"} variant="brand">
+                      2525
+                    </Chip>
```
why don't we support arbitrary ports, like on self-hosted?

fraserdrops replied:
It's a requirement that they set up a secure SMTP server, so we limit it to secure ports and security is required (in self-hosted there's a 'none' option. @nvoxland can probably explain it better than me but it's to do with our hosted AWS setup that we can't just connect to random SMTP servers (maybe SPAM related)

fraserdrops replied:
I am not an expert on this, so @nvoxland or @dpsutton please chime in here, but I believe these are the standard secure SMTP ports, and providers such as AWS don't like making connections to non-standard ports because spam

iethree commented on frontend/src/metabase/admin/settings/components/Email/CloudSMTPConnectionForm.tsx:
seems very strange to me to have a whole parallel form for essentially the same thing for cloud / self-hosted. couldn't we just have a little wrapper that provides the correct form values and sends the correct payload when saving based on whether cloud smtp is enabled?

fraserdrops replied:
> Essentially the same thing

40% of the fields are different due to the port and security fields being non-optional and being Chip components. Combined with yup schema also having slight subtle differences, it seemed messy to try and throw conditionals through a single form component

iethree commented on frontend/src/metabase/admin/settings/components/widgets/AdminSettingInput.tsx:
```diff
@@ -296,5 +296,5 @@ export function SetByEnvVarWrapper<SettingName extends SettingKey>({
     );
   }
-  return children;
+  return <div>{children}</div>;
```
why do we need this?

iethree reviewed (changes requested):
The biggest issue here is a lack of an enterprise plugin, and there's a few other things that could be nicer throughout.

I really find the way this lives in two sets of settings fundamentally very confusing. It's pretty counter-intuitive that in a cloud environment you can't change `smtp-host`, but the user-configurable one is `cloud-smtp-host` 🥴 . Wouldn't this all be much much simpler and require half the frontend code if the backend just stored the cloud-provided credentials somewhere and allowed the frontend to just continue setting the `smtp-xx` settings? Even if the backend stores them in `cloud-smtp-xx` settings, we could just ignore those completely on the frontend, right?

fraserdrops commented on frontend/src/metabase/admin/settings/components/widgets/SendTestEmailWidget.tsx:
```diff
@@ -0,0 +1,55 @@
+          {getTestEmailErrorMessage(sendTestEmailResult.error)}
+        </Text>
+      )}
+      <Button onClick={handleSendTestEmail} disabled={isLoading}>
+        {isLoading ? t`Sending...` : t`Send test email`}
+      </Button>
```
Agreed. I've asked for design feedback but not had any yet

fraserdrops commented on frontend/src/metabase/admin/settings/components/widgets/SendTestEmailWidget.tsx:
```diff
@@ -0,0 +1,55 @@
+import { useSetting, useToast } from "metabase/common/hooks";
+import { Button, Text } from "metabase/ui";
+
+export function SendTestEmailWidget() {
```
Here is the [Slack thread](https://metaboat.slack.com/archives/C090M4R2ELV/p1750354228521719) for the full context. Part of the assumption was that our current 'Test connection' check that runs when you save should catch errors like that, so I would be interested in reproducing that to see if we can catch it at the save stage cc @nvoxland 

TLDR; with the cloud smtp flow, Send Test Email in the form is more confusing because it's testing a configuration that might not be currently enabled. Also, because the modal auto-closes after save it would be easy to miss that the button even exists.


> The current user flow is a bit confusing, because the 'Test email button' is only active if a configuration has been added, but the configuration modal automatically closes.
What currently happens (self-hosted)
> 1. User opens smtp config modal
> 2. User creates config, saves it
> 3. Modal closes
> 4. Next time modal is open, Send test email is available
> 5. If a user clears the config the button is hidden again
> To me it makes sense to have this test button on the main email page, because it's only active after the configuration is complete

> I think this is confusing because it makes it look like the test email is coming from the active configuration. I believe we should only send test emails from whatever the active smtp is, rather than letting them send a test email from a non-active configuration
> Otherwise it looks like it's already fully set up, but they might not have actually activated the correct config

fraserdrops commented on frontend/src/metabase/api/email.ts:
```diff
@@ -20,6 +23,18 @@ export const settingsApi = Api.injectEndpoints({
       invalidatesTags: (_, error) =>
         invalidateTags(error, [tag("session-properties")]),
     }),
+    updateCloudEmailSMTPSettings: builder.mutation<
+      void,
+      CloudEmailSMTPSettings
+    >({
+      query: (cloudEmailSettings) => ({
+        method: "PUT",
+        url: `/api/email/cloud`,
+        body: cloudEmailSettings,
+      }),
+      invalidatesTags: (_, error) =>
+        invalidateTags(error, [tag("session-properties")]),
+    }),
```
Sure. What determines if it should be an `/ee/` route or not?

iethree replied:
I'm not entirely sure, but probably that it's a route only accessible on ee instances

fraserdrops commented:
> I really find the way this lives in two sets of settings fundamentally very confusing. It's pretty counter-intuitive that in a cloud environment you can't change `smtp-host`, but the user-configurable one is `cloud-smtp-host` 🥴 . Wouldn't this all be much much simpler and require half the frontend code if the backend just stored the cloud-provided credentials somewhere and allowed the frontend to just continue setting the `smtp-xx` settings? Even if the backend stores them in `cloud-smtp-xx` settings, we could just ignore those completely on the frontend, right?

@iethree A key bit of context you may or may not have is that the initial `smtp-xx` settings we set on the cloud instance need to be preserved because the user can switch between the `Manged` and `Cloud Custom` configurations. So the backend at least needs to make a distinction, although it is still possible to hide this from the frontend.

The question for me is if hiding from the frontend and switching out lots of stuff on the backend is a worthwhile tradeoff:
1. Changes the mental model of settings - one setting === one value is no longer true (although, was it ever true)?
2. There's actually quite a lot of differences - different components, different validation on the api, different form validation. Seems like a good chance hiding all of this will not be for free. Just for example, if there was a new requirement that we should show the user the `Managed` config settings whlie `Custom SMTP` is active, we wouldn't be able to support this if we're hiding the difference from the frontend. 

@nvoxland could you provide your perspective from the BE PoV?

fraserdrops commented on frontend/src/metabase/forms/components/FormChipGroup/FormChipGroup.tsx:
```diff
@@ -26,7 +26,6 @@ export const FormChipGroup = forwardRef(function FormChipGroup(
 ) {
   const [{ value }, _meta, { setValue }] = useField(name);
-
   const handleChange = useCallback(
```
Pumpin those numbers

fraserdrops commented:
> What if we just change it so that the `cloud-smtp-xx` settings are the ones that hold the non-user-editable settings that are set by cloud. I think that makes things much easier. Then, the `smtp-xx` are always user-defined and we can use the same form on the frontend, and conditionally have a `requireSecure` prop or something when we're in a cloud environment. We could also use the same API for saving, etc. We'd also have to put much less code in the enterprise plugins - probably just the radio toggle thingy

Are there any API/compatibility issues with moving the settings around? Otherwise I'm pretty for this suggestion.

Only nit is that the form still needs to be different (whether it's different components or some conditionals within it) because of the security we discussed elsewhere

nvoxland commented:
> > I really find the way this lives in two sets of settings fundamentally very confusing. It's pretty counter-intuitive that in a cloud environment you can't change `smtp-host`, but the user-configurable one is `cloud-smtp-host` 🥴 . Wouldn't this all be much much simpler and require half the frontend code if the backend just stored the cloud-provided credentials somewhere and allowed the frontend to just continue setting the `smtp-xx` settings? Even if the backend stores them in `cloud-smtp-xx` settings, we could just ignore those completely on the frontend, right?
> 
> @iethree A key bit of context you may or may not have is that the initial `smtp-xx` settings we set on the cloud instance need to be preserved because the user can switch between the `Manged` and `Cloud Custom` configurations. So the backend at least needs to make a distinction, although it is still possible to hide this from the frontend.
> 
> The question for me is if hiding from the frontend and switching out lots of stuff on the backend is a worthwhile tradeoff:
> 
> 1. Changes the mental model of settings - one setting === one value is no longer true (although, was it ever true)?
> 2. There's actually quite a lot of differences - different components, different validation on the api, different form validation. Seems like a good chance hiding all of this will not be for free. Just for example, if there was a new requirement that we should show the user the `Managed` config settings whlie `Custom SMTP` is active, we wouldn't be able to support this if we're hiding the difference from the frontend.
> 
> @nvoxland could you provide your perspective from the BE PoV?

I think partly you're hitting up against the naming of the settings which we were also thinking is overly confusing, but were going to leave as-is until things are working and then re-ask what name would make more sense and fix up the code if we have a better one. I thought we had a linear ticket for that discussion, but I don't see it? Something like `override-smtp-*` maybe would help vs. `cloud-smtp-*` ? Or, your suggestion of flipping around the "cloud-*" as the base version and smtp- as the "override" works well too.

As far as hiding the overridden vs. actual settings from the frontend: I think it would be fine to do in the backend if that would make it easier, or keep it separate if that would make it easier. 

If we wanted to hide it, I'm thinking it would be like this:

1. Change the `PUT /api/email` API:
- Take arguments of just `host`, `port`, etc. instead of `smtp-host`, etc. so it's more generic in naming
- Add an argument of "active-server" that takes "standard" or "override" (or whatever naming we want).
- Add arguments for the from-name and from-address
- Make it be a "POST" not a "PUT" so you can just modify changed portions without assuming it's specifying everything, allowing you to more easily change just "from-name" or "active-server"

2. Create a `GET /api/email` API:
- Returns the currently configurable settings (if hosted pro, return the "override" values)
- Have to say what settings are locked by env variables, maybe as a list in the response?
- Could return a flag like "requireSecure" to drive the FE logic

Then the FE would call those two APIs for everything. If you call `/api/settings` you will see the raw/actual settings. But you'd generally not want to use those directly from the frontend.

fraserdrops committed:
Add comment to tests

fraserdrops committed:
Cloud >override naming (#60375)

* cloud- -> override-

this is more in line with what we are thinking: it's an override. and it
just so happens to be in cloud. but cloud sets the default so it gets
confusing if the override is called cloud because cloud doesn't set it,
this holds the user override of cloud

---------

Co-authored-by: dan sutton <dan@dpsutton.com>

fraserdrops commented on frontend/src/metabase/admin/settings/components/widgets/EmailFromAddressWidget.tsx:
```diff
@@ -0,0 +1,59 @@
+        name="cloud-email-from-address"
+      />
+    );
+  }
```
On the block simplification specifically: I'm not sure I'm a fan of conditionally changing the name here, for two reasons:
1. Conceptually, I prefer to think of rendering a totally different Input rather than swapping out internals of the same one, which relates to:
2. It relies on the implementation of AdminSettingInput to avoid flickering or having a transient stale value as it loads the new setting key

In practice, will this be an issue? Probably not, because the feature flags are not going to change out of nowhere. Still not a habit I think we should get into

Also I don't think the snippet captures all the conditions correctly, I still would need an outer conditional to check if it's either `!isHosted || isCloudSMTPEnabled` because if it isn't either of those I render a different component

fraserdrops committed:
Merge branch 'master' into feature-cloud-smtp-whitelabeling

fraserdrops committed:
Fix radio indicator styling issues

nvoxland committed:
Moved new EE apis to /api/ee

nvoxland committed:
Fixed failing tests

nvoxland committed:
Fixed failing tests

fraserdrops committed:
Text changes

nvoxland committed:
Fixed failing tests

nvoxland committed:
Merge remote-tracking branch 'origin/feature-cloud-smtp-whitelabeling' into feature-cloud-smtp-whitelabeling

fraserdrops committed:
Fix unit test

fraserdrops committed:
Extract smtp enterprise code (#60477)

* Extract enterprise code

* Export smtp-override plugin and add tests

fraserdrops commented:
@iethree I believe I've addressed the main concerns from the previous comments in https://github.com/metabase/metabase/pull/60477

iethree commented on frontend/src/metabase/admin/settings/components/Email/BaseSMTPConnectionForm.tsx:
```diff
@@ -0,0 +1,332 @@
+          message: t`Email settings updated`,
+        });
+        onClose();
+      } catch (error) {
+        sendToast({
+          icon: "warning",
+          toastColor: "error",
+          message: getErrorMessage(error, t`Error updating email settings`),
+        });
+
+        // throw to allow the form to handle the error
+        throw error;
```
I think this catch is preventing the line item errors from showing up in the UI?
<img width="428" height="110" alt="Screenshot 2025-07-10 at 4 52 13 PM" src="https://github.com/user-attachments/assets/8e93fc2a-0f13-4ead-a982-cbad4a702761" />

edpaget reviewed (approved):
Looks good to me, just want to plus-one Ryan's point of the feature name casing.

fraserdrops committed:
Map between settings and form keys (#60813)

fraserdrops committed:
Use div component to prevent console warnings

fraserdrops committed:
Rename token feature to use underscores

fraserdrops committed:
Use actual radios instead of radio.card

fraserdrops committed:
Merge branch 'master' into feature-cloud-smtp-whitelabeling

fraserdrops committed:
Fix typo

fraserdrops committed:
Fix cypress test

nvoxland committed:
Update src/metabase/channel/email.clj

Co-authored-by: Edward Paget <edpaget@users.noreply.github.com>
